patches/api/0441-Expose-native-map-methods-in-PDC.patch

```diff
@@ -0,0 +1,288 @@
+From: Xander <xander.dekeijzer@student.hu.nl>
+Co-authored-by: TheRealRyGuy <ryan@ryguy.me>
+Date: Tue, 3 Oct 2023 07:28:50 +0200
+Subject: [PATCH] Expose native map methods in PDC
+
+
+diff --git a/src/main/java/org/bukkit/persistence/PersistentDataContainer.java b/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
+index 57609b7793122e135fa0c3b926500849379637b2..b221b97fd540b1f16aea3d6fce72d8a09c7a5885 100644
+--- a/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
++++ b/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
+@@ -9,7 +9,7 @@ import org.jetbrains.annotations.Nullable;
+  * This interface represents a map like object, capable of storing custom tags
+  * in it.
+  */
+-public interface PersistentDataContainer {
++public interface PersistentDataContainer extends Iterable<net.kyori.adventure.key.Key> {
+ 
+     /**
+      * Stores a metadata value on the {@link PersistentDataHolder} instance.
+@@ -153,9 +153,247 @@ public interface PersistentDataContainer {
+     PersistentDataAdapterContext getAdapterContext();
+ 
+     // Paper start
++    /**
++     * Returns an iterator over elements of type {@link net.kyori.adventure.key.Key}.
++     *
++     * @return an Iterator bound to the set of keys from {@link #keySet()}
++     */
++    @Override
++    default @NotNull java.util.Iterator<net.kyori.adventure.key.Key> iterator() {
++        return this.keySet().iterator();
++    }
++
++    /**
++     * Get a set of {@link net.kyori.adventure.key.Key}s present on this {@link PersistentDataContainer}
++     * instance.
++     * <p>
++     * Any changes made to the returned set will not be reflected on the
++     * instance.
++     *
++     * @return the key set
++     */
++    @NotNull
++    Set<net.kyori.adventure.key.Key> keySet();
++
++    /**
++     * Stores a metadata value on the {@link PersistentDataHolder} instance.
++     * <p>
++     * This API cannot be used to manipulate minecraft data, as the values will
++     * be stored using your namespace. This method will not override any existing
++     * value the {@link PersistentDataHolder} may have stored under the provided
++     * key.
++     *
++     * @param key   the key this value will be stored under
++     * @param type  the type this tag uses
++     * @param value the value stored in the tag
++     * @param <T>   the generic java type of the tag value
++     * @param <Z>   the generic type of the object to store
++     *
++     * @throws NullPointerException     if the key is null
++     * @throws NullPointerException     if the type is null
++     * @throws NullPointerException     if the value is null
++     * @throws IllegalArgumentException if no suitable adapter will be found for
++     *                                  the {@link PersistentDataType#getPrimitiveType()}
++     */
++    <T, Z> void setIfAbsent(@NotNull net.kyori.adventure.key.Key key, @NotNull PersistentDataType<T, Z> type, @NotNull Z value);
```

Comment:
This isn't a Map method and I think everything it does is covered by `computeIfAbsent` and `Functions.constant()`.

Reply:
I think the previous author chose this name because PDC uses set() instead of put() but I'm not sure

Reply:
I think if we are moving most of this to me as map like as possible, I think it would be best to use put here.

Reply:
So should we change the set method to put then to, or add put as the adventure key method to not break api?

Reply:
So:
`set(NamespacedKey, ...)`
`put(Key, ...)`
`put(String, ...)`?

Reply:
We certainly cannot rename the method and break API.
I think it is honestly fine to leave it as put.
A "mistake" made when naming it, but certainly not remotely bad enough to warrent a rename.

But if we add new stuff, I think it is a good idea to stick to the map like naming if that is what the new methods aim to achieve.